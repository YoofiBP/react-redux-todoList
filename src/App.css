li {
    list-style-type: none;
}

body {
    align-content: center;
}

.input-group {
    width: 50%;
}