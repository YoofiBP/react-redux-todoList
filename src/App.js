import React from "react";
import { Welcome } from "./Components/InputBox";
import DateComponent from './Components/DateComponent'
import ToDoList from './Components/ToDoList';
import InputBox from './Components/InputBox';
import "./App.css";

/*
App Array of Items Lives here
  Date Renders Based on Array of Items
  InputBoxContainer Renders based on Form Text
  TodoListContainer Renders Based on Array of Items
    TodoListItem

State:
Array of Items
Form Text
Date

================To dos==================
Implement DRY principle where applicable
Add comments and use descriptive names
Organization > Optimisation
Read popular source code
Take time to think about the problem before
Make this all much better
ÃŸ
Add move To tomorrow button
*/

export class App extends React.Component {
  constructor(props) {
    super(props);
    /** App State = todoItems (Array), currentDate (String) */
    this.state = {
      userId: this.props.userId,
    };
  }


  render() {
    let uid = 2;
    return (
      <>
      <div className="row justify-content-center">
      <Welcome userId={uid}/>
      </div>
        <div className="row justify-content-center">
          <DateComponent/>
        </div>
        <div className="row justify-content-center">
          <InputBox />
        </div>
        <div className="row justify-content-center">
          <div className="mx-auto col-5 border pt-2">
            <ToDoList/>
            <div style={{ textAlign: "center" }}>
              <button id="logout" className="btn btn-outline-secondary" onClick={this.props.logOut}>Logout</button>
            </div>
          </div>
        </div>
      </>
    );
  }
}
