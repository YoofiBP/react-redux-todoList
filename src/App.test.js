import React from 'react';
import { create } from 'react-test-renderer';
import { shallow} from 'enzyme';
import { DateContainer, Welcome } from "./Containers/DateContainer";
import { ToDoListContainer } from "./Containers/TodoListContainer";
import { InputBoxContainer } from "./Containers/InputBoxContainer";
import {App} from './App';
import dayjs from "dayjs";
import { spy } from 'sinon';

it('renders without crashing', () => {
    shallow(<App/>);
})

it('Matches snapshot', () => {
        const app = create(<App />)
        let appJSON = app.toJSON();
        expect(appJSON).toMatchSnapshot();
})

it('Logout button executes prop when clicked', () => {
    const logoutSpy = spy();
    const wrapper = shallow(<App logOut={logoutSpy}/>)
    const logoutButton = wrapper.find('#logout');
    logoutButton.simulate('click');

    expect(logoutSpy.called).toEqual(true);
})

it("renders all children components", () => {
    const wrapper = shallow(<App />);
    expect(wrapper.containsAllMatchingElements([Welcome, DateContainer, ToDoListContainer, InputBoxContainer])).toEqual(true);
    expect(wrapper.find('#logout')).toHaveLength(1);
})

it("Has default state", () => {
    const wrapper = shallow(<App userId={1}/>);
    expect(wrapper.state('userId')).toEqual(expect.any(Number));
    expect(wrapper.state('todoItems')).toBeInstanceOf(Array);
    expect(wrapper.state('todoItems')[0]).toEqual(expect.objectContaining({
        date: expect.any(String),
        tasks: expect.any(Array)
    }))
    expect(wrapper.state('todoItems')[0].tasks[0]).toEqual(expect.objectContaining({
        userId: expect.any(Number),
        id: expect.any(Number),
        title: expect.any(String),
        completed: expect.any(Boolean)
    }));
    expect(wrapper.state('currentDate')).toEqual(expect.any(String))
})

it('Function to change date works', () => {
    const wrapper = shallow(<App />);
    const date = dayjs("05/03/1996").format("MM/DD/YYYY");
    wrapper.setState({currentDate: date});
    expect(wrapper.state('currentDate')).toEqual("05/03/1996");
    wrapper.instance().changeDate("back");
    expect(wrapper.state('currentDate')).toEqual("05/02/1996");
    wrapper.instance().changeDate("forward");
    expect(wrapper.state('currentDate')).toEqual("05/03/1996");
})

describe("Add, Delete, Move, Check Testing", () => {
    const wrapper = shallow(<App/>)
    let idx, title;

    beforeEach(() => {
    const date = dayjs("05/03/1996").format("MM/DD/YYYY");
    wrapper.setState({currentDate: date});
    wrapper.instance().changeItem('add', 'React testing with Jest');
    idx = wrapper.state('todoItems').filter(day => day.date === "05/03/1996")[0].tasks[0].id;
    title = wrapper.state('todoItems').filter(day => day.date === "05/03/1996")[0].tasks[0].title;
    })

    afterEach(() => {
        wrapper.instance().changeItem('delete', idx)
    })

    it("Add item functionality working", () => {
        expect(wrapper.state('todoItems')).toEqual(expect.arrayContaining([expect.objectContaining({
            date: "05/03/1996",
            tasks: expect.arrayContaining([{
            userId: expect.any(Number),
            id: expect.any(Number),
            title: "React testing with Jest",
            completed: false
            }])
        })]))
    })
    
    it("Delete functionality working", () => {
        wrapper.instance().changeItem('delete', idx);
        expect(wrapper.state('todoItems')).toEqual(expect.arrayContaining([expect.objectContaining({
            date: "05/03/1996",
            tasks: expect.not.arrayContaining([{
                userId: expect.any(Number),
                id: idx,
                title: "React testing with Jest",
                completed: false
            }])
        })]))
        })
    
        //REVISIT
    it("Move Item functionality working", () => {
        wrapper.instance().changeItem('move', idx+"/"+title);
        expect(wrapper.state('currentDate')).toEqual('05/04/1996')
        expect(wrapper.state('todoItems')).toEqual(expect.arrayContaining([expect.objectContaining({
            date: "05/03/1996",
            tasks: expect.not.arrayContaining([{
                userId: expect.any(Number),
                id: expect.any(Number),
                title: expect.any(String),
                completed: expect.any(Boolean)
            }])
        })]))
    })
        
    it("Check Item functionality working", () => {
        wrapper.instance().changeItem('check', idx);
        expect(wrapper.state('todoItems')).toEqual(expect.arrayContaining([expect.objectContaining({
            date: "05/03/1996",
            tasks: expect.arrayContaining([{
                userId: expect.any(Number),
                id: idx,
                title: "React testing with Jest",
                completed: true
            }])
        })]))
})

    it("Adds multiple items correctly", () => {
        wrapper.instance().changeItem('add', 'React testing with Jest 2');
        wrapper.instance().changeItem('add', 'React testing with Jest 3');
        wrapper.instance().changeItem('add', 'React testing with Jest 4');
        expect(wrapper.state('todoItems').filter(day => day.date === "05/03/1996")[0].tasks).toHaveLength(4);
    })
})

describe("Children have appropriate props", () => {
    const wrapper = shallow(<App userId={1}/>);
    let dateBackward, dateForward, date, changeItem, todoItems;

    beforeEach(() => {
        date = "05/03/1996";
        wrapper.setState({currentDate: date});
        dateForward = dateBackward = wrapper.instance().changeDate;
        changeItem = wrapper.instance().changeItem;
        todoItems = wrapper.state('todoItems').filter(day => day.date === date);
    })

    it("Welcome", () => {
        const welcomeWrapper = wrapper.find(Welcome);
        expect(welcomeWrapper.props()).toEqual(expect.objectContaining({
            userId: 1
        }))
    })

    it("DateContainer", () => {
        const dateContainerWrapper = wrapper.find(DateContainer);
        expect(dateContainerWrapper.props()).toEqual(expect.objectContaining({
            date: date,
            dateForward: dateForward,
            dateBackward: dateBackward
        }))
    })

    it("InputBoxContainer", () => {
        const inputBoxContainerWrapper = wrapper.find(InputBoxContainer);
        expect(inputBoxContainerWrapper.props()).toEqual(expect.objectContaining({
            onClick: changeItem
        }));
    })

    it("TodoListContainer", () => {
        const toDoListContainerWrapper = wrapper.find(ToDoListContainer);
        expect(toDoListContainerWrapper.props()).toEqual(expect.objectContaining({
            todoItems: todoItems,
            onClick: changeItem,
            onChange: changeItem,
            onMove: changeItem
        }))
    })
})

describe("Functions are binding properly", () => {
    const wrapper = shallow(<App userId={1}/>);
    const date = "05/03/1996";
    
    beforeEach(() => {
        wrapper.setState({currentDate: date});
    })
    it("DateContainer", () => {
        const dateContainerWrapper = wrapper.find(DateContainer);
        dateContainerWrapper.prop('dateForward')('forward');
        expect(wrapper.state('currentDate')).toEqual("05/04/1996");
        dateContainerWrapper.prop('dateForward')('back');
        expect(wrapper.state('currentDate')).toEqual("05/03/1996");
    })

    it("InputBoxContainer", () => {
        const inputBoxWrapper = wrapper.find(InputBoxContainer);
        inputBoxWrapper.prop('onClick')('add', "React Test 1");
        expect(wrapper.state('todoItems')).toEqual(expect.arrayContaining([expect.objectContaining({
            date: "05/03/1996",
            tasks: expect.arrayContaining([{
            userId: expect.any(Number),
            id: expect.any(Number),
            title: "React Test 1",
            completed: false
            }])
        })]))
    })

    //FINISH TESTING BINDING OF METHODS TO CHILD COMPONENTS
    it("ToDoListContainer", () => {
        const todoListWrapper = wrapper.find(ToDoListContainer);
        wrapper.setState({currentDate: "03/17/2020"});
        todoListWrapper.prop('onClick')('delete', 1);
        expect(wrapper.state('todoItems')).toEqual(expect.arrayContaining([expect.objectContaining({
            date: "03/17/2020",
            tasks: expect.not.arrayContaining([{
                userId: 1,
                id: 1,
                title: "Learn React",
                completed:false
            }])
        })]))

        todoListWrapper.prop('onMove')('move', "2/Conduct Internal Vapt");
        expect(wrapper.state('todoItems')).toEqual(expect.arrayContaining([expect.objectContaining({
            date: "03/17/2020",
            tasks: expect.not.arrayContaining([{
                userId: 1,
                id: 2,
                title: "Conduct Internal VAPT",
                completed: false
            }])
        })]))

        wrapper.setState({currentDate: "03/17/2020"});
        todoListWrapper.prop('onChange')('check', 3);
        expect(wrapper.state('todoItems')).toEqual(expect.arrayContaining([expect.objectContaining({
            date: "03/17/2020",
            tasks: expect.arrayContaining([{
                userId: expect.any(Number),
                id: 3,
                title: expect.any(String),
                completed: true
            }])
        })]))
    })
    
})

