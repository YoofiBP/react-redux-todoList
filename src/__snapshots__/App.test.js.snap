// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshot 1`] = `
Array [
  <div
    className="row justify-content-center"
  >
    <h1>
      Welcome, 
      Yoofi
      !
    </h1>
  </div>,
  <div
    className="row justify-content-center"
  >
    <button
      aria-label="delete"
      className="MuiButtonBase-root MuiIconButton-root p-2 MuiIconButton-sizeSmall MuiIconButton-edgeEnd"
      disabled={false}
      id="back"
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M11.67 3.87L9.9 2.1 0 12l9.9 9.9 1.77-1.77L3.54 12z"
          />
        </svg>
      </span>
    </button>
    <h1>
      Today
    </h1>
    <button
      aria-label="delete"
      className="MuiButtonBase-root MuiIconButton-root p-2 MuiIconButton-sizeSmall MuiIconButton-edgeEnd"
      disabled={false}
      id="forward"
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M5.88 4.12L13.76 12l-7.88 7.88L8 22l10-10L8 2z"
          />
        </svg>
      </span>
    </button>
  </div>,
  <div
    className="row justify-content-center"
  >
    <div
      className="input-group mb-2"
    >
      <form
        className="input-group mb-2 w-100"
        onSubmit={[Function]}
      >
        <input
          aria-describedby="button-addon2"
          aria-label="New Item"
          className="form-control"
          id="title"
          onChange={[Function]}
          placeholder="New Item"
          type="text"
          value=""
        />
        <input
          className="btn btn-outline-secondary"
          type="submit"
          value="Add Item"
        />
      </form>
    </div>
  </div>,
  <div
    className="row justify-content-center"
  >
    <div
      className="mx-auto col-5 border pt-2"
    >
      <div
        style={
          Object {
            "textAlign": "center",
          }
        }
      >
        <button
          className="btn btn-outline-secondary"
          id="logout"
        >
          Logout
        </button>
      </div>
    </div>
  </div>,
]
`;
